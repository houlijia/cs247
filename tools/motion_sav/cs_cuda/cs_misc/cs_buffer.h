#ifndef __CS_BUFFER_H__
#define __CS_BUFFER_H__

struct cs_buf_desc {
	int size ;	// number of bytes per buffer
	int unit_size ; // size in int and float ;
	int cnt ;	// number of such buffers
} ;

struct cs_buf {
	char *d_A ; // device address
	struct cs_buf *np ;
} ;

struct cs_buf_list {
	struct cs_buf *bp ;
	int cnt ;

	int max_used ;
	int used ;
} ;

int cs_buffer_init ( struct cs_buf_desc *cbp, int cnt, int lock ) ;

// for debug purpose

void p_buffer_dbg( const char * ) ;

char * cs_get_free_list ( int idx ) ;
void cs_put_free_list ( char *bp, int idx ) ;

void cs_buf_swap (float **fp1, float **fp2 ) ;

#endif 
