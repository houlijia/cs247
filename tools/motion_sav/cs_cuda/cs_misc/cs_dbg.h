#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
