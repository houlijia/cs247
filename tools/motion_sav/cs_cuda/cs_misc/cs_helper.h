#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
