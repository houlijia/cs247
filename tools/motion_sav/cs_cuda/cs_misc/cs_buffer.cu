#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>
#include <semaphore.h>

#include "cs_cuda.h"
#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_buffer.h"

// #define CUDA_DBG
// #define CUDA_DBG1

static int cs_buf_cnt ;
static int cs_buf_free_cnt = 0 ;

static struct cs_buf *cs_buf_free ;
static struct cs_buf_list *cs_buf_lp ;

static sem_t mutex ;
static int do_lock = 0 ;

void cs_put_free_buf ( struct cs_buf *bp ) ;
struct cs_buf * cs_get_free_buf () ;


// lock : 1->do semaphore for the critical section 
int
cs_buffer_init ( struct cs_buf_desc *cbp, int cnt, int lock )
{
	char *d_p ;
	int icnt, total, i, j ;
	struct cs_buf_desc *cp ;
	struct cs_buf *bp ;

	do_lock = lock ;

	if ( do_lock )
	{
		if ( sem_init ( &mutex, 0, 1 ) < 0 )
		{
			printf("%s : sem_init failed \n", __func__ ) ;
			return ( 0 ) ;
		}
	}

	cs_buf_cnt = cnt ;
	cp = cbp ;

	total = 0 ;
	icnt = 0 ;
	for ( i = 0 ; i < cnt ; i++ )
	{
		total += (((( cp->size - 1 ) >> 2 ) + 1 ) << 2 ) * cp->cnt ;  

		icnt += cp->cnt ;

#ifdef CUDA_DBG 
		printf("%s: i %d total %d size %d cnt %d total cnt %d\n", __func__, i,
			total, cp->size, cp->cnt, icnt ) ;
#endif 
		cp++ ;
	}

	if (( i = cudaMalloc ( &d_p, total )) != cudaSuccess )
	{
		printf("%s: malloc failed %d \n", __func__, i ) ;
		return ( 0 ) ;
	}

	if (!( cs_buf_lp = ( struct cs_buf_list *)malloc ( sizeof ( 
		struct cs_buf_list ) * cs_buf_cnt )))
	{
#ifdef CUDA_DBG 
		printf("%s: cs_list failed \n", __func__ ) ;
#endif 
		return ( 0 ) ;
	}

	if (!( bp = cs_buf_free = ( struct cs_buf *)malloc ( sizeof ( struct cs_buf ) * icnt )))
	{
#ifdef CUDA_DBG 
		printf("%s: cs_bufp failed \n", __func__ ) ;
#endif 
		return ( 0 ) ;
	}

	cs_buf_free = NULL ;

	while ( icnt-- )
	{
		cs_put_free_buf ( bp ) ;
		bp++ ;
	}

	cp = cbp ;
	for ( i = 0 ; i < cs_buf_cnt ; i++ )
	{
		cs_buf_lp[i].bp = NULL ;
		cs_buf_lp[i].cnt = 0 ;

		total = (((( cp->size - 1 ) >> 2 ) + 1 ) << 2 ) ;  
		for ( j = 0 ; j < cp->cnt ; j++ )
		{
			cs_put_free_list( d_p, i ) ;
			d_p += total ;
		}

		cs_buf_lp[i].used = 0 ;
		cs_buf_lp[i].max_used = 0 ;
			
		cp++ ;
	}

	p_buffer_dbg("init") ;

	return ( 1 ) ;
}

void
p_buf_free(const char *s )
{
	struct cs_buf *cb ;
	int i = 0 ;

	printf("%s: %s cnt %d \n", __func__, s, cs_buf_free_cnt ) ;

	if ( do_lock )
		sem_wait ( &mutex ) ;

	cb = cs_buf_free ;
	while ( cb )
	{
		i++ ;
		printf("%i :: %p \n", i, cb ) ;
		cb = cb->np ;
	}

	if ( do_lock )
		sem_post( &mutex ) ;
}

void
p_buf_list( const char *s )
{
	struct cs_buf *cb ;
	int i, j, k ;

#ifdef CUDA_DBG 
	printf("%s: %s cs_buf_cnt %d\n", __func__, s, cs_buf_cnt ) ;
#endif

	if ( do_lock )
		sem_wait ( &mutex ) ;

	for ( i = 0 ; i < cs_buf_cnt ; i++ )
	{
		cb = cs_buf_lp[i].bp ;
		j = cs_buf_lp[i].cnt ;

		printf("%s : idx %d used %d cnt %d max_used %d \n",
			__func__, i, cs_buf_lp[i].used, cs_buf_lp[i].cnt, cs_buf_lp[i].max_used ) ;
		
		for ( k = 0 ; k < j ; k++ )
		{
#ifdef CUDA_DBG 
			printf("i %d k %d cb %p d_A %p np %p \n", i, k, cb, cb->d_A, cb->np ) ;
#endif 
			cb = cb->np ;
		}
	}	

	if ( do_lock )
		sem_post( &mutex ) ;
}

void
p_buffer_dbg( const char *s )
{
#ifdef CUDA_DBG 
	printf("--------------------------------------------------------------- B\n") ;
	p_buf_free( s ) ;
	p_buf_list( s ) ;
	printf("--------------------------------------------------------------- E\n") ;
#endif 
}

char *
cs_get_free_list ( int idx )
{
	struct cs_buf *cs_bufp ;
	char *bp = NULL ;

#ifdef CUDA_DBG 
	printf("%s: idx %d \n", __func__, idx ) ;
#endif 

	if ( idx < 0 || idx >= cs_buf_cnt )
	{
#ifdef CUDA_DBG 
		printf("%s: wrong idx %d \n", __func__, idx ) ;
#endif 
		return ( NULL ) ;
	}

	if ( do_lock )
		sem_wait ( &mutex ) ;

	cs_bufp = cs_buf_lp[ idx ].bp ;

	if ( cs_bufp )
	{
		cs_buf_lp[ idx ].bp = cs_bufp->np ; 

		bp = cs_bufp->d_A ;

		cs_buf_lp[ idx ].cnt-- ;
		cs_buf_lp[ idx ].used++ ;

		if ( cs_buf_lp[ idx ].used > cs_buf_lp[ idx ].max_used )
			cs_buf_lp[ idx ].max_used++ ;
	} 

#ifdef CUDA_DBG 
	printf("%s: idx %d bp %p cnt %d\n", __func__, idx, bp, cs_buf_lp[idx].cnt ) ;
#endif 

	if ( do_lock )
		sem_post( &mutex ) ;

	if ( cs_bufp )
		cs_put_free_buf( cs_bufp ) ;

	if ( bp == NULL )
	{
		printf("buffer ERR : idx %d OUT OF BUFFER wait forever\n", idx) ;
		while ( 1 )
			sleep ( 60 ) ;
	}

#ifdef CUDA_OBS 
	p_buffer_dbg("cs_get_free_list") ;
#endif 
	return ( bp ) ;
}

void
cs_put_free_list ( char *bp, int idx )
{
	struct cs_buf *fbp ;

#ifdef CUDA_DBG 
	printf("%s: bp %p idx %d \n", __func__, bp, idx ) ;
#endif 

	if ( idx < 0 || idx >= cs_buf_cnt )
	{
#ifdef CUDA_DBG 
		printf("%s: wrong idx %d \n", __func__, idx ) ;
#endif 
		return ;
	}

	fbp = cs_get_free_buf() ;

	if ( fbp == NULL )
	{
		printf("%s: no free buf left idx %d bp %p\n", __func__, idx, bp ) ;
		return ;
	}

	if ( do_lock )
		sem_wait ( &mutex ) ;

	fbp->d_A = bp ;

	if ( cs_buf_lp[ idx ].bp )
		fbp->np = cs_buf_lp[ idx ].bp ;
	else
		fbp->np = NULL ;

	cs_buf_lp[ idx ].bp = fbp ;
	cs_buf_lp[ idx ].cnt++ ;
	cs_buf_lp[ idx ].used-- ;

#ifdef CUDA_DBG 
	printf("%s: bp %p d_p %p idx %d cnt %d\n", __func__, cs_buf_lp[idx].bp,
		cs_buf_lp[idx].bp->d_A, idx, cs_buf_lp[idx].cnt ) ;
#endif 

	if ( do_lock )
		sem_post( &mutex ) ;
}

struct cs_buf *
cs_get_free_buf ()
{
	struct cs_buf *bp = NULL ;

	if ( do_lock )
		sem_wait ( &mutex ) ;

	if ( cs_buf_free )
	{
		bp = cs_buf_free ;
		cs_buf_free = cs_buf_free->np ;
		cs_buf_free_cnt-- ;
	}

#ifdef CUDA_DBG 
	printf("%s: bp %d cnt %d \n", __func__, bp, cs_buf_free_cnt ) ;
#endif 

	if ( do_lock )
		sem_post( &mutex ) ;

	return ( bp ) ;
}

void
cs_put_free_buf ( struct cs_buf *bp )
{
	if ( do_lock )
		sem_wait ( &mutex ) ;

	bp->np = cs_buf_free ;
	cs_buf_free = bp ;
	cs_buf_free_cnt++ ;

#ifdef CUDA_DBG 
	printf("%s: bp %d cnt %d \n", __func__, bp, cs_buf_free_cnt ) ;
#endif 

	if ( do_lock )
		sem_post( &mutex ) ;
}

void
cs_buf_swap (float **fp1, float **fp2 )
{
	float *fp ;

	fp = *fp1 ;
	*fp1 = *fp2 ;
	*fp2 = fp ;
}

