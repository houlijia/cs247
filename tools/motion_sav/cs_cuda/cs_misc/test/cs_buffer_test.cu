#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_buffer.h"

#define CUDA_DBG
#define BUF_SIZE_INT		10

float *fp1 ;

struct cs_buf_desc cbd[] = {
	{ 1000, 2 },
	{ 2000, 3 },
	{ 3000, 4 }
} ;

int
main( int ac, char *av[] )
{
	char *p1, *p2, *p3, *p4 ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;

	cs_buffer_init ( cbd, 3 ) ;

	p_buffer_dbg("after init") ;

	printf("get 2 from 0 -----------------------------------------------------\n") ;

	p1 = cs_get_free_list ( 0 ) ;
	p2 = cs_get_free_list ( 0 ) ;

	printf("get 2 from 0: p1 %p p2 %p \n", p1, p2 ) ;

	p_buffer_dbg("get 2 from 0") ;

	cs_put_free_list( p2, 0 ) ;
	cs_put_free_list( p1, 0 ) ;

	p_buffer_dbg("put 2 from 0") ;

	printf("get from 1/2 -----------------------------------------------------\n") ;

	p1 = cs_get_free_list ( 1 ) ;
	p2 = cs_get_free_list ( 2 ) ;

	printf("get from 1/2: p1 %p p2 %p \n", p1, p2 ) ;

	p_buffer_dbg("get from 1/2") ;

	cs_put_free_list( p1, 1 ) ;
	cs_put_free_list( p2, 2 ) ;

	p_buffer_dbg("put to 1/2") ;

	printf("get 4 from 1 -----------------------------------------------------\n") ;

	p1 = cs_get_free_list ( 1 ) ;
	p2 = cs_get_free_list ( 1 ) ;
	p3 = cs_get_free_list ( 1 ) ;
	p4 = cs_get_free_list ( 1 ) ;

	printf("get 4 from 1: p1 %p p2 %p p3 %p p4 %p \n", p1, p2, p3, p4 ) ;

	p_buffer_dbg("get from 1/2") ;

	if ( p1 )
		cs_put_free_list( p2, 1 ) ;

	if ( p2 ) 
		cs_put_free_list( p3, 1 ) ;

	if ( p3 )
		cs_put_free_list( p1, 1 ) ;

	if ( p4 )
		cs_put_free_list( p4, 1 ) ;

	p_buffer_dbg("put 4 to 1") ;
}
