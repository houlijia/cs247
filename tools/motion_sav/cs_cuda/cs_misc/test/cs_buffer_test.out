cs_buffer_init: i 0 total 2000 size 1000 cnt 2 total cnt 2
cs_buffer_init: i 1 total 8000 size 2000 cnt 3 total cnt 5
cs_buffer_init: i 2 total 20000 size 3000 cnt 4 total cnt 9
icnt 9
cs_put_free_buf: bp 20215280 cnt 1 
cs_put_free_buf: bp 20215296 cnt 2 
cs_put_free_buf: bp 20215312 cnt 3 
cs_put_free_buf: bp 20215328 cnt 4 
cs_put_free_buf: bp 20215344 cnt 5 
cs_put_free_buf: bp 20215360 cnt 6 
cs_put_free_buf: bp 20215376 cnt 7 
cs_put_free_buf: bp 20215392 cnt 8 
cs_put_free_buf: bp 20215408 cnt 9 
cs_put_free_list: bp 0x1300200000 idx 0 
cs_get_free_buf: bp 20215408 cnt 8 
cs_put_free_list: bp 0x1347670 d_p 0x1300200000 idx 0 cnt 1
cs_put_free_list: bp 0x13002003e8 idx 0 
cs_get_free_buf: bp 20215392 cnt 7 
cs_put_free_list: bp 0x1347660 d_p 0x13002003e8 idx 0 cnt 2
cs_put_free_list: bp 0x13002007d0 idx 1 
cs_get_free_buf: bp 20215376 cnt 6 
cs_put_free_list: bp 0x1347650 d_p 0x13002007d0 idx 1 cnt 1
cs_put_free_list: bp 0x1300200fa0 idx 1 
cs_get_free_buf: bp 20215360 cnt 5 
cs_put_free_list: bp 0x1347640 d_p 0x1300200fa0 idx 1 cnt 2
cs_put_free_list: bp 0x1300201770 idx 1 
cs_get_free_buf: bp 20215344 cnt 4 
cs_put_free_list: bp 0x1347630 d_p 0x1300201770 idx 1 cnt 3
cs_put_free_list: bp 0x1300201f40 idx 2 
cs_get_free_buf: bp 20215328 cnt 3 
cs_put_free_list: bp 0x1347620 d_p 0x1300201f40 idx 2 cnt 1
cs_put_free_list: bp 0x1300202af8 idx 2 
cs_get_free_buf: bp 20215312 cnt 2 
cs_put_free_list: bp 0x1347610 d_p 0x1300202af8 idx 2 cnt 2
cs_put_free_list: bp 0x13002036b0 idx 2 
cs_get_free_buf: bp 20215296 cnt 1 
cs_put_free_list: bp 0x1347600 d_p 0x13002036b0 idx 2 cnt 3
cs_put_free_list: bp 0x1300204268 idx 2 
cs_get_free_buf: bp 20215280 cnt 0 
cs_put_free_list: bp 0x13475f0 d_p 0x1300204268 idx 2 cnt 4
--------------------------------------------------------------- B
p_buf_free: init cnt 0 
p_buf_list: init cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x1347630 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x13475f0 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
--------------------------------------------------------------- B
p_buf_free: after init cnt 0 
p_buf_list: after init cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x1347630 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x13475f0 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
get 2 from 0 -----------------------------------------------------
cs_get_free_list: idx 0 
cs_put_free_buf: bp 20215392 cnt 1 
cs_get_free_list: idx 0 bp 0x13002003e8 cnt 1
cs_get_free_list: idx 0 
cs_put_free_buf: bp 20215408 cnt 2 
cs_get_free_list: idx 0 bp 0x1300200000 cnt 0
get 2 from 0: p1 0x13002003e8 p2 0x1300200000 
--------------------------------------------------------------- B
p_buf_free: get 2 from 0 cnt 2 
1 :: 0x1347670 
2 :: 0x1347660 
p_buf_list: get 2 from 0 cs_buf_cnt 3
i 1 k 0 cb 0x1347630 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x13475f0 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
cs_put_free_list: bp 0x1300200000 idx 0 
cs_get_free_buf: bp 20215408 cnt 1 
cs_put_free_list: bp 0x1347670 d_p 0x1300200000 idx 0 cnt 1
cs_put_free_list: bp 0x13002003e8 idx 0 
cs_get_free_buf: bp 20215392 cnt 0 
cs_put_free_list: bp 0x1347660 d_p 0x13002003e8 idx 0 cnt 2
--------------------------------------------------------------- B
p_buf_free: put 2 from 0 cnt 0 
p_buf_list: put 2 from 0 cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x1347630 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x13475f0 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
get from 1/2 -----------------------------------------------------
cs_get_free_list: idx 1 
cs_put_free_buf: bp 20215344 cnt 1 
cs_get_free_list: idx 1 bp 0x1300201770 cnt 2
cs_get_free_list: idx 2 
cs_put_free_buf: bp 20215280 cnt 2 
cs_get_free_list: idx 2 bp 0x1300204268 cnt 3
get from 1/2: p1 0x1300201770 p2 0x1300204268 
--------------------------------------------------------------- B
p_buf_free: get from 1/2 cnt 2 
1 :: 0x13475f0 
2 :: 0x1347630 
p_buf_list: get from 1/2 cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 1 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 1 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 2 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
cs_put_free_list: bp 0x1300201770 idx 1 
cs_get_free_buf: bp 20215280 cnt 1 
cs_put_free_list: bp 0x13475f0 d_p 0x1300201770 idx 1 cnt 3
cs_put_free_list: bp 0x1300204268 idx 2 
cs_get_free_buf: bp 20215344 cnt 0 
cs_put_free_list: bp 0x1347630 d_p 0x1300204268 idx 2 cnt 4
--------------------------------------------------------------- B
p_buf_free: put to 1/2 cnt 0 
p_buf_list: put to 1/2 cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x13475f0 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x1300200fa0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x13002007d0 np (nil) 
i 2 k 0 cb 0x1347630 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
get 4 from 1 -----------------------------------------------------
cs_get_free_list: idx 1 
cs_put_free_buf: bp 20215280 cnt 1 
cs_get_free_list: idx 1 bp 0x1300201770 cnt 2
cs_get_free_list: idx 1 
cs_put_free_buf: bp 20215360 cnt 2 
cs_get_free_list: idx 1 bp 0x1300200fa0 cnt 1
cs_get_free_list: idx 1 
cs_put_free_buf: bp 20215376 cnt 3 
cs_get_free_list: idx 1 bp 0x13002007d0 cnt 0
cs_get_free_list: idx 1 
cs_get_free_list: idx 1 bp (nil) cnt 0
get 4 from 1: p1 0x1300201770 p2 0x1300200fa0 p3 0x13002007d0 p4 (nil) 
--------------------------------------------------------------- B
p_buf_free: get from 1/2 cnt 3 
1 :: 0x1347650 
2 :: 0x1347640 
3 :: 0x13475f0 
p_buf_list: get from 1/2 cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 2 k 0 cb 0x1347630 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
cs_put_free_list: bp 0x1300200fa0 idx 1 
cs_get_free_buf: bp 20215376 cnt 2 
cs_put_free_list: bp 0x1347650 d_p 0x1300200fa0 idx 1 cnt 1
cs_put_free_list: bp 0x13002007d0 idx 1 
cs_get_free_buf: bp 20215360 cnt 1 
cs_put_free_list: bp 0x1347640 d_p 0x13002007d0 idx 1 cnt 2
cs_put_free_list: bp 0x1300201770 idx 1 
cs_get_free_buf: bp 20215280 cnt 0 
cs_put_free_list: bp 0x13475f0 d_p 0x1300201770 idx 1 cnt 3
--------------------------------------------------------------- B
p_buf_free: put 4 to 1 cnt 0 
p_buf_list: put 4 to 1 cs_buf_cnt 3
i 0 k 0 cb 0x1347660 d_A 0x13002003e8 np 0x1347670 
i 0 k 1 cb 0x1347670 d_A 0x1300200000 np (nil) 
i 1 k 0 cb 0x13475f0 d_A 0x1300201770 np 0x1347640 
i 1 k 1 cb 0x1347640 d_A 0x13002007d0 np 0x1347650 
i 1 k 2 cb 0x1347650 d_A 0x1300200fa0 np (nil) 
i 2 k 0 cb 0x1347630 d_A 0x1300204268 np 0x1347600 
i 2 k 1 cb 0x1347600 d_A 0x13002036b0 np 0x1347610 
i 2 k 2 cb 0x1347610 d_A 0x1300202af8 np 0x1347620 
i 2 k 3 cb 0x1347620 d_A 0x1300201f40 np (nil) 
--------------------------------------------------------------- E
